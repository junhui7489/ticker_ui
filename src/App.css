.App {
  text-align: center;
}

.App-logo {
  height: 40vmin;
  pointer-events: none;
}

@media (prefers-reduced-motion: no-preference) {
  .App-logo {
    animation: App-logo-spin infinite 20s linear;
  }
}

.App-header {
  background-color: #282c34;
  min-height: 100vh;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  font-size: calc(10px + 2vmin);
  color: white;
}

.App-link {
  color: #61dafb;
}

@keyframes App-logo-spin {
  from {
    transform: rotate(0deg);
  }
  to {
    transform: rotate(360deg);
  }
}

.appbar {
  min-width:100%;
  padding-top: 0px;
  position:fixed;
  top:0%;
  left:0%;
  display: flex;
}

.appbar > .logo {
  width: 10%;
}

.listTable tbody tr td {
  font-size: 15px;
  color: blueviolet;
  font-family: sans-serif;
}

.listTable{
  border-collapse: collapse;
}

.listTable tr{
margin-bottom: 10px;
}

.listTable tr:nth-child(even){
  background-color: #a6edb3;
}

.listTable thead{
  background-color:	#c5dde3;
}

.listTable th {
  font-size: 15px;
  color: purple;
  font-family: sans-serif;
  font-weight: bolder;
}

.listTable tbody tr:hover td, .listTable tbody tr:hover th {
  background-color:lightgray;
  cursor: pointer;
}