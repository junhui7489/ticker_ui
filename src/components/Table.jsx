export default function Table(props){
  const data = props.data;
  const tickerName = props.tickerName;
  const sourceName = props.sourceName;
  
  if(data.length > 5){
  return(
  <table className="listTable">
    <thead>
    <tr>
      <th>Time</th>
      <th>Price</th>
    </tr>
    </thead>
    <tbody>
      {data.filter((row) => row.stock == tickerName && row.source == sourceName).slice(-5).reverse().map((row,i) => 
      <tr key={i}>
        <td>{row.time}</td>
        <td>{row.price}</td>
      </tr>
      ) 
      }
  </tbody>
  </table>
  )
  }
  else{
    return(
    <table>
    <tbody>
      <tr>
        <th>Time</th>
        <th>Price</th>
      </tr>
        {data.filter((row) => row.stock == tickerName && row.source == sourceName).reverse().map((row,i) => 
        <tr key={i}>
          <td>{row.time}</td>
          <td>{row.price}</td>
        </tr>
        ) 
        }
    </tbody>
    </table>
    )
    }
}