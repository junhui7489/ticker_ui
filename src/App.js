import logo from './logo.svg';
import React, {useState,useContext, useEffect} from 'react';
import './App.css';
import ButtonAppBar from './components/Appbar';
import '../node_modules/bootstrap/dist/css/bootstrap.min.css';
import { Source1Context } from './contexts/Source1Context';
import { Source3Context } from './contexts/Source3Context';
import { Source2Context } from './contexts/Source2Context';
import { useTickerContext } from './contexts/useTickerContext';
import { useSourcesContext } from './contexts/useSourcesContext';
import Table from './components/Table';

function App(props) {
  const [list, setList] = React.useState([]); //list points to table data
  const [source1, setSource1] = useContext(Source1Context);
  const [source2, setSource2] = useContext(Source2Context);
  const [source3, setSource3] = useContext(Source3Context);
  const [getTickers] = useContext(useTickerContext);
  const [getSources] = useContext(useSourcesContext);
  const [currentSource, setCurrentSource] = React.useState("SRC1");
  const [currentTicker, setCurrentTicker] = React.useState(["goog"]);

  useEffect(()=>{
    if (currentSource == "SRC1"){
      console.log(source1);
    setList(source1);
    }
    else if (currentSource == "SRC2"){
      setList(source2);
    }
    else{
      setList(source3);
    }
  },[source1, source2, source3, currentSource, currentTicker])

  const changeSource = (e) => {
    setCurrentSource(e.target.value)
  }

  const changeTicker = (e) =>{
    setCurrentTicker(e.target.value)
  }

  return (
    <div>
      <ButtonAppBar/>
      <br/>
      <div className="container">
        <h2>Ticker</h2>
        <div className="col text-center">
            <div className="dropdown">
            <label for="cars">Price Source:&nbsp;</label>
            <select name="sources" id="sources" onChange={changeSource}>
              {
                getSources.map((source) => <option value={source}>{source}</option>)
              }
            </select>
            </div>
      <br/>
        </div>
            <div className="col text-center">
            <div className="dropdown">
            <label for="cars">Current Ticker:&nbsp;</label>
            <select name="sources" id="sources" onChange={changeTicker}>
              {
                getTickers.map((ticker) => <option value={ticker}>{ticker}</option>)
              }
            </select>
            </div>
        </div>
        <br/>
        <br/>
        <div className='row'>
          <Table data={list} sourceName={currentSource} tickerName={currentTicker}></Table>
        </div>
      </div>
    </div>
  );
}

export default App;
