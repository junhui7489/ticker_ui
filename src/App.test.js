import { render, screen, waitFor } from '@testing-library/react';
import { act } from "react-dom/test-utils";
import { unmountComponentAtNode } from "react-dom";
import App from './App';
import ButtonAppBar from './components/Appbar';
import Table from './components/Table';
import mockData from './test/mockData.json'
import source1Data from './test/source1Data.json';
import source2Data from './test/source2Data.json';
import source3Data from './test/source3Data.json';
import testSourcesData from './test/testSources.json';
import testTickersData from './test/testTickers';
import  Source1Provider  from './contexts/Source1Context';
import Source2Provider from './contexts/Source2Context';
import Source3Provider from './contexts/Source3Context';
import UseSourcesProvider from './contexts/useSourcesContext';
import UseTickerProvider from './contexts/useTickerContext';
import index from './index';


describe('Application bar', ()=>{
  test('Logo text', ()=> {
    render(
      <ButtonAppBar/>,
    )
    const linkElement = screen.getByText("LOGO");
    expect(linkElement).toBeInTheDocument();
  })

  test('Picture', ()=>{
    render(
      <ButtonAppBar/>,
    )
    const logo = screen.getByRole('img');
    expect(logo).toHaveAttribute('src', 'logo-social.png');
    expect(logo).toHaveAttribute('alt', 'Logo');
  })
})


describe('Table', ()=> {
  test('Time text', ()=> {
    const tickerName = "goog";
    const sourceName = "SRC1";
    let test = mockData
    render(
      <Table data={test} sourceName={sourceName} tickerName={tickerName}></Table>,
    )
    const linkElement = screen.getByText("Time");
    expect(linkElement).toBeInTheDocument();
  })

  test('Price text', ()=> {
    const tickerName = "goog";
    const sourceName = "SRC1";
    let test = mockData
    render(
      <Table data={test} sourceName={sourceName} tickerName={tickerName}></Table>,
    )
    const linkElement = screen.getByText("Price");
    expect(linkElement).toBeInTheDocument();
  })

  test('if data length less than or equal to 5', ()=> {
    const tickerName = "goog";
    const sourceName = "SRC1";
    let test = mockData.slice(-5);
    render(
      <Table data={test} sourceName={sourceName} tickerName={tickerName}></Table>,
    )
    const linkElement = screen.getByText("Price");
    expect(linkElement).toBeInTheDocument();
  })

})

describe('Main Page test', ()=> {  
  let container = null;
  beforeEach(() => {
  // setup a DOM element as a render target
  container = document.createElement("div");
  document.body.appendChild(container);
});

afterEach(() => {
  // cleanup on exiting
  unmountComponentAtNode(container);
  container.remove();
  container = null;
});
  test('context snapshot testing', () => {
    let test = mockData
    jest.spyOn(global, "fetch").mockImplementation(() =>
    Promise.resolve({
      json: () => Promise.resolve(test)
    })
    );
    act(() => {
      render(      
      <UseSourcesProvider>
      <UseTickerProvider>
        <Source1Provider>
          <Source2Provider>
            <Source3Provider>
            <App/>
            </Source3Provider>
          </Source2Provider>
        </Source1Provider>
      </UseTickerProvider>
    </UseSourcesProvider>, container
        )
      
      // render components
    });


    const linkElement = screen.getAllByText("Ticker");
    expect(linkElement).toMatchSnapshot();
  });

  test('Price Source text exists snapshot testing', () => {
    let test = mockData
    jest.spyOn(global, "fetch").mockImplementation(() =>
    Promise.resolve({
      json: () => Promise.resolve(test)
    })
    );
    act(() => {
      render(      
      <UseSourcesProvider>
      <UseTickerProvider>
        <Source1Provider>
          <Source2Provider>
            <Source3Provider>
            <App/>
            </Source3Provider>
          </Source2Provider>
        </Source1Provider>
      </UseTickerProvider>
    </UseSourcesProvider>, container
        )
      
      // render components
    });
    const linkElement = screen.getAllByText("Price Source:");
    expect(linkElement).toMatchSnapshot();
  });

  test('dropdown option exists snapshot testing', () => {
    let test = mockData;
    jest.spyOn(global, "fetch").mockImplementation(() =>
    Promise.resolve({
      json: () => Promise.resolve(test)
    })
  );

  act(() => {
    render(      
    <UseSourcesProvider>
    <UseTickerProvider>
      <Source1Provider>
        <Source2Provider>
          <Source3Provider>
          <App/>
          </Source3Provider>
        </Source2Provider>
      </Source1Provider>
    </UseTickerProvider>
  </UseSourcesProvider>, container
      )
    
    // render components
  });
    const dropdownButton = screen.getAllByRole('combobox')
    
    expect(dropdownButton.length).toEqual(2) // it exist
  });

  it('Ticker text exists or not', async() => {

    const source1Context = source1Data['updated_src1'];
    const source2Context = source2Data['updated_src2'];
    const source3Context = source3Data['updated_src3'];
    const getTickersContext = testTickersData['list'];
    const getSourcesContext = testSourcesData['list'];
    jest.spyOn(global, "fetch").mockImplementation(() =>
    Promise.resolve({
      json: () => Promise.resolve(getSourcesContext)
    })
    );
    jest.spyOn(global, "fetch").mockImplementation(() =>
    Promise.resolve({
      json: () => Promise.resolve(getTickersContext)
    })
    );

    render(
        <UseSourcesProvider value={getSourcesContext}>
        <UseTickerProvider value={getTickersContext}>
          <Source1Provider value={source1Context}>
            <Source2Provider value={source2Context}>
              <Source3Provider value={source3Context}>
              <App/>
              </Source3Provider>
            </Source2Provider>
          </Source1Provider>
        </UseTickerProvider>
      </UseSourcesProvider>, container
    )

    await waitFor(() => {
      const element = screen.getByText("Ticker")
  
      expect(element).toBeInTheDocument();
    });
    
  });

})


describe('root test', ()=> {  
  it("Should render app without crashing", () => {
    expect(
        JSON.stringify(
          Object.assign({}, index, { _reactInternalInstance: 'censored' }),
        ),
      ).toMatchSnapshot();
});

})
