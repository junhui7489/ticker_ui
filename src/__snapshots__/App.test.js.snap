// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main Page test Price Source text exists snapshot testing 1`] = `
Array [
  <label
    for="cars"
  >
    Price Source:Â 
  </label>,
]
`;

exports[`Main Page test context snapshot testing 1`] = `
Array [
  <h2>
    Ticker
  </h2>,
]
`;

exports[`root test Should render app without crashing 1`] = `"{\\"_reactInternalInstance\\":\\"censored\\"}"`;
